hack/e2e.go

```diff
@@ -67,125 +60,198 @@ func appendError(errs []error, err error) []error {
 	return errs
 }
 
-func main() {
-	log.SetFlags(log.LstdFlags | log.Lshortfile)
-	flag.Parse()
-
+func validWorkingDirectory() error {
 	cwd, err := os.Getwd()
 	if err != nil {
-		log.Fatalf("Could not get pwd: %v", err)
+		return fmt.Errorf("could not get pwd: %v", err)
 	}
 	acwd, err := filepath.Abs(cwd)
 	if err != nil {
-		log.Fatalf("Failed to convert to an absolute path: %v", err)
+		return fmt.Errorf("failed to convert %s to an absolute path: %v", cwd, err)
 	}
+	// This also matches "kubernetes_skew" for upgrades.
 	if !strings.Contains(filepath.Base(acwd), "kubernetes") {
-		// TODO(fejta): cd up into  the kubernetes directory
-		log.Fatalf("Must run from kubernetes directory: %v", cwd)
+		return fmt.Errorf("must run from kubernetes directory root: %v", acwd)
+	}
+	return nil
+}
+
+type TestCase struct {
+	XMLName   xml.Name `xml:"testcase"`
+	ClassName string   `xml:"classname,attr"`
+	Name      string   `xml:"name,attr"`
+	Time      float64  `xml:"time,attr"`
+	Failure   string   `xml:"failure,omitempty"`
+}
+
+type TestSuite struct {
+	XMLName  xml.Name `xml:"testsuite"`
+	Failures int      `xml:"failures,attr"`
+	Tests    int      `xml:"tests,attr"`
+	Time     float64  `xml:"time,attr"`
+	Cases    []TestCase
+}
+
+var suite TestSuite
+
+func xmlWrap(name string, f func() error) error {
+	start := time.Now()
+	err := f()
+	duration := time.Since(start)
+	c := TestCase{
+		Name:      name,
+		ClassName: "e2e.go",
+		Time:      duration.Seconds(),
+	}
+	if err != nil {
+		c.Failure = err.Error()
+		suite.Failures++
+	}
+	suite.Cases = append(suite.Cases, c)
+	suite.Tests++
+	return err
```

Comment:
Got it because you're just returning the error (that's what you cannot change, creation of the error)